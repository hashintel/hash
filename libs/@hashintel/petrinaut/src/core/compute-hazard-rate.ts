import { enumerateWeightedMarkingIndices } from "./enumerate-weighted-markings";

export type TransitionTokenCombination<TToken> = {
  placeId: string;
  tokenIndices: number[];
  tokens: TToken[];
};

/**
 * Compute the hazard rate (Î») for every combination of tokens that can enable a transition.
 *
 * A transition may require multiple tokens per input place (or even zero). For each place we:
 *   1. Enumerate every unique combination of tokens that satisfies its weight (requirement).
 *   2. Take the Cartesian product across all input places, giving every possible joint marking.
 *   3. Invoke `lambda` with the concrete token data for that joint marking.
 *
 * The caller provides the available tokens per place, the number of tokens that must be selected
 * (weight), and a user supplied `lambda` function that evaluates the hazard for a specific
 * combination of tokens.
 *
 * @returns One entry per joint combination, preserving the ordering returned by
 *          `enumerateWeightedMarkingIndices`.
 */
export function computeHazardRate<TToken>({
  inputs,
  lambda,
}: {
  inputs: Array<TransitionInputTokens<TToken>>;
  lambda: HazardRateLambda<TToken>;
}): Array<HazardRateResult<TToken>> {
  if (inputs.length === 0) {
    return [
      {
        combination: [],
        hazardRate: lambda({ combination: [] }),
      },
    ];
  }

  const placeSpecs = inputs.map((input) => ({
    tokenCount: input.tokens.length,
    weight: input.weight,
  }));

  const enumerated = enumerateWeightedMarkingIndices(placeSpecs);

  if (enumerated.length === 0) {
    return [];
  }

  // Pre-compute offsets into the flattened index combinations so we can slice quickly per place.
  const offsets: number[] = [];
  let runningOffset = 0;
  for (const input of inputs) {
    offsets.push(runningOffset);
    runningOffset += input.weight;
  }

  return enumerated.map((flatCombination) => {
    const combination = inputs.map<TransitionTokenCombination<TToken>>(
      (input, index) => {
        const start = offsets[index] ?? 0;
        const end = start + input.weight;
        const tokenIndices = flatCombination.slice(start, end);
        const tokens = tokenIndices.map(
          (tokenIndex) => input.tokens[tokenIndex]!,
        );

        return {
          placeId: input.placeId,
          tokenIndices,
          tokens,
        };
      },
    );

    const hazardRate = lambda({ combination });

    return {
      combination,
      hazardRate,
    };
  });
}
