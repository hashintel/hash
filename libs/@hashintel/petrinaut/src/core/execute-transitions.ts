import { computePossibleTransition } from "./compute-possible-transition";
import { removeTokensFromSimulationFrame } from "./remove-tokens-from-simulation-frame";
import type { ID, SimulationFrame } from "./types";

type PlaceID = ID;

/**
 * Adds tokens to multiple places in the simulation frame.
 *
 * Takes a SimulationFrame and a Map of Place IDs to arrays of token values,
 * and returns a new SimulationFrame with:
 * - The specified tokens added to each place's section in the buffer
 * - Each place's count incremented by the number of added tokens
 * - All subsequent places' offsets adjusted accordingly
 *
 * @param frame - The simulation frame to modify
 * @param tokensToAdd - Map from Place ID to array of token values to add (each token is an array of numbers)
 * @returns A new SimulationFrame with the tokens added
 * @throws Error if a place is not found or token dimensions don't match
 */
function addTokensToSimulationFrame(
  frame: SimulationFrame,
  tokensToAdd: Map<PlaceID, number[][]>
): SimulationFrame {
  // If no tokens to add, return frame as-is
  if (tokensToAdd.size === 0) {
    return frame;
  }

  // Validate all places and token dimensions first
  for (const [placeId, tokens] of tokensToAdd) {
    const placeState = frame.places.get(placeId);
    if (!placeState) {
      throw new Error(`Place with ID ${placeId} not found in simulation frame.`);
    }

    // Validate that all tokens have the correct dimensions
    const expectedDimensions = placeState.instance.dimensions;
    for (const token of tokens) {
      if (token.length !== expectedDimensions) {
        throw new Error(
          `Token dimension mismatch for place ${placeId}. Expected ${expectedDimensions}, got ${token.length}.`
        );
      }
    }
  }

  // Calculate total size increase needed in buffer
  let totalSizeIncrease = 0;
  for (const [placeId, tokens] of tokensToAdd) {
    const placeState = frame.places.get(placeId)!;
    const tokenSize = placeState.instance.dimensions;
    totalSizeIncrease += tokens.length * tokenSize;
  }

  // Create a new buffer with increased size
  const newBuffer = new Float64Array(frame.buffer.length + totalSizeIncrease);

  // Process places in order of their offsets to build the new buffer
  const placesByOffset = Array.from(frame.places.entries()).sort(
    (a, b) => a[1].offset - b[1].offset
  );

  const newPlaces = new Map(frame.places);
  let sourceIndex = 0;
  let targetIndex = 0;

  for (const [placeId, placeState] of placesByOffset) {
    const { count, instance } = placeState;
    const tokenSize = instance.dimensions;
    const placeSize = count * tokenSize;

    // Copy existing tokens from this place
    for (let i = 0; i < placeSize; i++) {
      newBuffer[targetIndex++] = frame.buffer[sourceIndex++]!;
    }

    // Add new tokens for this place if any
    const newTokens = tokensToAdd.get(placeId);
    if (newTokens) {
      for (const token of newTokens) {
        for (const value of token) {
          newBuffer[targetIndex++] = value;
        }
      }

      // Update this place's count
      newPlaces.set(placeId, {
        ...placeState,
        count: count + newTokens.length,
      });
    }
  }

  // Recalculate all offsets based on the new buffer layout
  let currentOffset = 0;
  for (const [placeId, _placeState] of placesByOffset) {
    const updatedState = newPlaces.get(placeId)!;
    const tokenSize = updatedState.instance.dimensions;
    const placeSize = updatedState.count * tokenSize;

    newPlaces.set(placeId, {
      ...updatedState,
      offset: currentOffset,
    });

    currentOffset += placeSize;
  }

  return {
    ...frame,
    buffer: newBuffer,
    places: newPlaces,
  };
}

/**
 * Executes all transitions sequentially on a simulation frame.
 *
 * This function:
 * 1. Iterates through all transitions in the frame
 * 2. For each transition, computes if it can fire using computePossibleTransition
 * 3. Immediately removes tokens after each transition fires (so subsequent transitions see the updated state)
 * 4. Accumulates all token additions across all transitions
 * 5. At the end, adds all accumulated tokens at once
 *
 * @param frame - The simulation frame to execute transitions on
 * @returns A new SimulationFrame with all transitions applied, or the original frame if no transitions fired
 */
export function executeTransitions(frame: SimulationFrame): SimulationFrame {
  // Map to accumulate all tokens to add: PlaceID -> array of token values
  const tokensToAdd = new Map<PlaceID, number[][]>();

  // Keep track of which transitions fired for updating timeSinceLastFiring
  const transitionsFired = new Set<ID>();

  // Start with the current frame and update it as transitions fire
  let currentFrame = frame;

  // Iterate through all transitions in the frame
  for (const [transitionId, _transitionState] of currentFrame.transitions) {
    // Compute if this transition can fire based on the current state
    const result = computePossibleTransition(currentFrame, transitionId);

    if (result !== null) {
      // Transition fired!
      transitionsFired.add(transitionId);

      // Immediately remove tokens from the current frame
      // Convert the result.remove Record to a Map
      const tokensToRemove = new Map<PlaceID, Set<number>>(
        Object.entries(result.remove)
      );
      currentFrame = removeTokensFromSimulationFrame(
        currentFrame,
        tokensToRemove
      );

      // Accumulate tokens to add
      for (const [placeId, tokenValues] of Object.entries(result.add)) {
        if (!tokensToAdd.has(placeId)) {
          tokensToAdd.set(placeId, []);
        }
        const existingTokens = tokensToAdd.get(placeId)!;
        existingTokens.push(...tokenValues);
      }
    }
  }

  // If no transitions fired, return the original frame
  if (transitionsFired.size === 0) {
    return frame;
  }

  // Add all new tokens at once
  const newFrame = addTokensToSimulationFrame(currentFrame, tokensToAdd);

  // Update transition timeSinceLastFiring
  const newTransitions = new Map(newFrame.transitions);
  for (const [transitionId, transitionState] of newFrame.transitions) {
    if (transitionsFired.has(transitionId)) {
      // Reset time since last firing for transitions that fired
      newTransitions.set(transitionId, {
        ...transitionState,
        timeSinceLastFiring: 0,
      });
    } else {
      // Increment time for transitions that didn't fire
      newTransitions.set(transitionId, {
        ...transitionState,
        timeSinceLastFiring:
          transitionState.timeSinceLastFiring + frame.simulation.dt,
      });
    }
  }

  return {
    ...newFrame,
    transitions: newTransitions,
    time: frame.time + frame.simulation.dt,
  };
}
