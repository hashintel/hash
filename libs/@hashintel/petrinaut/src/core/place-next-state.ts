export type ODEType = "euler" | "midpoint" | "rk4";

export type PlaceDifferentialEquation = (
  currentState: Float64Array,
  dimensions: number,
  numberOfTokens: number,
) => Float64Array;

/**
 * Takes current Place state, a differential equation defining its dynamics,
 * an ODE solving method, and a time step (dt), and computes the next state
 * of the Place using the specified ODE method.
 *
 * Currently, only the Euler method is implemented.
 */
export function getPlaceNextState(
  placeState: Float64Array,
  dimensions: number,
  numberOfTokens: number,
  differentialEquation: PlaceDifferentialEquation,
  odeType: ODEType,
  dt: number,
): Float64Array {
  if (odeType !== "euler") {
    throw new Error(
      `ODE type ${odeType} not implemented yet. Use Euler method.`,
    );
  }

  const derivatives = differentialEquation(
    placeState,
    dimensions,
    numberOfTokens,
  );

  return (
    placeState
      // Apply Euler method: nextState = currentState + derivative * dt
      .map((value, index) => value + derivatives[index]! * dt)
  );
}
