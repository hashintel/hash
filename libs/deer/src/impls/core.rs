mod array;
mod bool;
mod bytes;
mod cell;
mod cmp;
mod marker;
mod mem;
mod num;
mod ops;
mod option;
mod result;
mod string;
mod sync;
mod time;
mod tuples;
mod unit;
