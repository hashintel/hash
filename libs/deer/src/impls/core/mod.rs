mod array;
mod atomic;
mod bool;
mod floating;
mod integral;
mod non_zero;
mod string;
mod unit;
