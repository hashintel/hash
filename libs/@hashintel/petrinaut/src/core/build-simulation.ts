import { compileUserCode } from "./helpers/compile-user-code";
import type {
  DifferentialEquationFn,
  LambdaFn,
  SimulationFrame,
  SimulationInput,
  SimulationInstance,
  TransitionKernelFn,
} from "./types/simulation";

/**
 * Builds a simulation instance and its initial frame from simulation input.
 *
 * Takes a SimulationInput containing:
 * - SDCPN definition (places, transitions, and their code)
 * - Initial marking (token distribution across places)
 * - Random seed
 * - Time step (dt)
 *
 * Returns a SimulationFrame with:
 * - A SimulationInstance containing compiled user code functions
 * - Initial token distribution in a contiguous buffer
 * - All places and transitions initialized with proper state
 *
 * @param input - The simulation input configuration
 * @returns The initial simulation frame ready for execution
 * @throws Error if place IDs in initialMarking don't match places in SDCPN
 * @throws Error if token dimensions don't match place dimensions
 * @throws Error if user code fails to compile
 */
export function buildSimulation(input: SimulationInput): SimulationFrame {
  const { sdcpn, initialMarking, seed, dt } = input;

  // Build maps for quick lookup
  const placesMap = new Map(sdcpn.places.map((place) => [place.id, place]));
  const transitionsMap = new Map(
    sdcpn.transitions.map((transition) => [transition.id, transition])
  );

  // Validate that all places in initialMarking exist in SDCPN
  for (const placeId of initialMarking.keys()) {
    if (!placesMap.has(placeId)) {
      throw new Error(
        `Place with ID ${placeId} in initialMarking does not exist in SDCPN`
      );
    }
  }

  // Validate token dimensions match place dimensions
  for (const [placeId, marking] of initialMarking) {
    const place = placesMap.get(placeId)!;
    const expectedSize = place.dimensions * marking.count;
    if (marking.values.length !== expectedSize) {
      throw new Error(
        `Token dimension mismatch for place ${placeId}. Expected ${expectedSize} values (${place.dimensions} dimensions Ã— ${marking.count} tokens), got ${marking.values.length}`
      );
    }
  }

  // Compile all differential equation functions
  const differentialEquationFns = new Map<string, DifferentialEquationFn>();
  for (const place of sdcpn.places) {
    try {
      const fn = compileUserCode<[Float64Array, number]>(
        place.differentialEquationCode,
        "Dynamics"
      );
      differentialEquationFns.set(place.id, fn as DifferentialEquationFn);
    } catch (error) {
      throw new Error(
        `Failed to compile differential equation for place ${place.id}: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  // Compile all lambda functions
  const lambdaFns = new Map<string, LambdaFn>();
  for (const transition of sdcpn.transitions) {
    try {
      const fn = compileUserCode<[number[][][]]>(
        transition.lambdaCode,
        "Lambda"
      );
      lambdaFns.set(transition.id, fn as LambdaFn);
    } catch (error) {
      throw new Error(
        `Failed to compile lambda function for transition ${transition.id}: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  // Compile all transition kernel functions
  const transitionKernelFns = new Map<string, TransitionKernelFn>();
  for (const transition of sdcpn.transitions) {
    try {
      const fn = compileUserCode<[number[][][]]>(
        transition.transitionKernelCode,
        "TransitionKernel"
      );
      transitionKernelFns.set(transition.id, fn as TransitionKernelFn);
    } catch (error) {
      throw new Error(
        `Failed to compile transition kernel for transition ${transition.id}: ${
          error instanceof Error ? error.message : String(error)
        }`
      );
    }
  }

  // Calculate buffer size and build place states
  let bufferSize = 0;
  const placeStates = new Map<
    string,
    { instance: (typeof sdcpn.places)[0]; offset: number; count: number }
  >();

  // Process places in a consistent order (sorted by ID)
  const sortedPlaceIds = Array.from(placesMap.keys()).sort();

  for (const placeId of sortedPlaceIds) {
    const place = placesMap.get(placeId)!;
    const marking = initialMarking.get(placeId);
    const count = marking?.count ?? 0;

    placeStates.set(placeId, {
      instance: place,
      offset: bufferSize,
      count,
    });

    bufferSize += place.dimensions * count;
  }

  // Build the initial buffer with token values
  const buffer = new Float64Array(bufferSize);
  let bufferIndex = 0;

  for (const placeId of sortedPlaceIds) {
    const marking = initialMarking.get(placeId);
    if (marking && marking.count > 0) {
      for (let i = 0; i < marking.values.length; i++) {
        buffer[bufferIndex++] = marking.values[i]!;
      }
    }
  }

  // Initialize transition states
  const transitionStates = new Map(
    sdcpn.transitions.map((transition) => [
      transition.id,
      {
        instance: transition,
        timeSinceLastFiring: 0,
      },
    ])
  );

  // Create the simulation instance (without frames initially)
  const simulationInstance: SimulationInstance = {
    id: sdcpn.id,
    title: sdcpn.title,
    places: placesMap,
    transitions: transitionsMap,
    differentialEquationFns,
    lambdaFns,
    transitionKernelFns,
    dt,
    rngState: seed,
    frames: [], // Will be populated with the initial frame
    currentFrameNumber: 0,
  };

  // Create the initial frame
  const initialFrame: SimulationFrame = {
    simulation: simulationInstance,
    time: 0,
    places: placeStates,
    transitions: transitionStates,
    buffer,
  };

  // Add the initial frame to the simulation instance
  simulationInstance.frames.push(initialFrame);

  return initialFrame;
}
