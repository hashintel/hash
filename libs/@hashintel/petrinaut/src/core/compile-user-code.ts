import { transformSync } from "@babel/core";

/**
 * Strips TypeScript type annotations from code to make it executable JavaScript.
 * Uses Babel to properly parse and transform TypeScript code.
 *
 * @param code - TypeScript code with type annotations
 * @returns JavaScript code without type annotations
 */
function stripTypeAnnotations(code: string): string {
  const result = transformSync(code, {
    filename: "input.ts",
    plugins: [
      [
        "@babel/plugin-transform-typescript",
        {
          allowDeclareFields: true,
          onlyRemoveTypeImports: true,
          isTSX: false,
        },
      ],
    ],
  });
  return result?.code ?? "";
}

/**
 * Compiles TypeScript/JavaScript module code into an executable function.
 * Expects a JavaScript module with a default export using a specific constructor function.
 *
 * @param code - The TypeScript/JavaScript module code with a default export.
 *               Should follow the pattern: `export default ConstructorFn((a, b, c) => { ... })`
 *               Can include TypeScript type annotations which will be stripped.
 * @param constructorFnName - The name of the constructor function that wraps the user code.
 *                            This must match the function used in the default export.
 * @returns A compiled function that can be executed
 *
 * @example
 * ```typescript
 * const code = `
 *   export default Dynamics((a: number, b: number, c: number) => {
 *     return a + b + c;
 *   });
 * `;
 * const fn = compileUserCode(code, "Dynamics");
 * const result = fn(1, 2, 3); // returns 6
 * ```
 */
export function compileUserCode<T extends unknown[] = unknown[]>(
  code: string,
  constructorFnName: string,
): (...args: T) => unknown {
  // Strip TypeScript type annotations and remove leading/trailing whitespace
  const sanitizedCode = stripTypeAnnotations(code.trim());

  // Verify that the code has a default export
  const defaultExportRegex = /export\s+default\s+/;
  if (!defaultExportRegex.test(sanitizedCode)) {
    throw new Error(
      `Module must have a default export. Expected pattern: export default ${constructorFnName}(...)`,
    );
  }

  // Verify that the default export uses the specified constructor function
  const constructorPattern = new RegExp(
    `export\\s+default\\s+${constructorFnName}\\s*\\(`,
  );
  if (!constructorPattern.test(sanitizedCode)) {
    throw new Error(
      `Default export must use the constructor function '${constructorFnName}'. Expected pattern: export default ${constructorFnName}(...)`,
    );
  }

  try {
    // Create a mock constructor function that extracts the wrapped function
    const mockConstructor = `
      function ${constructorFnName}(fn) {
        if (typeof fn !== 'function') {
          throw new Error('${constructorFnName} expects a function as argument');
        }
        return fn;
      }
    `;

    // Create an executable module-like environment
    const executableCode = `
      ${mockConstructor}
      let __default_export__;
      ${sanitizedCode.replace(/export\s+default\s+/, "__default_export__ = ")}
      return __default_export__;
    `;

    // Use Function constructor to create and execute the module
    // eslint-disable-next-line @typescript-eslint/no-implied-eval, no-new-func, @typescript-eslint/no-unsafe-call
    const compiledFunction = new Function(executableCode)() as (
      ...args: T
    ) => unknown;

    if (typeof compiledFunction !== "function") {
      throw new Error(
        `Expected default export to be a function, got ${typeof compiledFunction}`,
      );
    }

    return compiledFunction;
  } catch (error) {
    throw new Error(
      `Failed to compile user code: ${
        error instanceof Error ? error.message : String(error)
      }`,
    );
  }
}
