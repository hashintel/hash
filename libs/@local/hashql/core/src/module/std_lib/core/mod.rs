use core::iter;

use super::{ItemDef, ModuleDef, StandardLibrary, StandardLibraryModule};
use crate::{
    heap::Heap,
    module::{item::IntrinsicValueItem, locals::TypeDef},
    symbol::Symbol,
};

pub(in crate::module::std_lib) mod bits;
pub(in crate::module::std_lib) mod bool;
pub(in crate::module::std_lib) mod cmp;
pub(in crate::module::std_lib) mod graph;
pub(in crate::module::std_lib) mod json;
pub(in crate::module::std_lib) mod math;
pub(in crate::module::std_lib) mod option;
pub(in crate::module::std_lib) mod result;
pub(in crate::module::std_lib) mod url;
pub(in crate::module::std_lib) mod uuid;

fn func<'heap>(
    lib: &StandardLibrary<'_, 'heap>,
    def: &mut ModuleDef<'heap>,

    name: &'static str,
    alias: &[&'static str],

    r#type: TypeDef<'heap>,
) {
    let value = IntrinsicValueItem { name, r#type };

    let ident = name.rsplit_once("::").expect("path should be non-empty").1;

    def.push_aliased(
        iter::once(ident)
            .chain(alias.iter().copied())
            .map(|name| lib.heap.intern_symbol(name)),
        ItemDef::intrinsic(value),
    );
}

pub(in crate::module::std_lib) struct Core;

impl<'heap> StandardLibraryModule<'heap> for Core {
    type Children = (
        self::bits::Bits,
        self::bool::Bool,
        self::cmp::Cmp,
        self::graph::Graph,
        self::json::Json,
        self::math::Math,
        self::option::Option,
        self::result::Result,
        self::url::Url,
        self::uuid::Uuid,
    );

    fn name(heap: &'heap Heap) -> Symbol<'heap> {
        heap.intern_symbol("core")
    }

    fn define(_: &mut StandardLibrary<'_, 'heap>) -> ModuleDef<'heap> {
        ModuleDef::new()
    }
}
