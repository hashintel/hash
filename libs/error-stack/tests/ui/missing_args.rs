#![expect(deprecated, reason = "`report!` is deprecated")]

use core::{error::Error, fmt};

use error_stack::{Report, bail, ensure, report};

#[derive(Debug)]
pub struct RootError;

impl fmt::Display for RootError {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt.write_str("root error")
    }
}

impl Error for RootError {}

fn empty_report_arg() -> Result<(), Report<RootError>> {
    let _ = report!();

    Ok(())
}

fn empty_bail_arg() -> Result<(), Report<RootError>> {
    bail!()
}

fn empty_ensure_arg() -> Result<(), Report<RootError>> {
    let _ = ensure!();

    Ok(())
}

fn missing_ensure_arg() -> Result<(), Report<RootError>> {
    let _ = ensure!(true);

    Ok(())
}

fn main() {}
