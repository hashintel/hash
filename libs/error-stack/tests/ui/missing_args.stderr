error: unexpected end of macro invocation
  --> tests/ui/missing_args.rs:17:5
   |
17 |     bail!()
   |     ^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$err:expr`
  --> src/macros.rs
   |
   |     ($err:expr $(,)?) => {{
   |      ^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/ui/missing_args.rs:21:13
   |
21 |     let _ = ensure!();
   |             ^^^^^^^^^ missing tokens in macro arguments
   |
note: while trying to match meta-variable `$cond:expr`
  --> src/macros.rs
   |
   |     ($cond:expr, $err:expr $(,)?) => {{
   |      ^^^^^^^^^^

error: unexpected end of macro invocation
  --> tests/ui/missing_args.rs:27:25
   |
27 |     let _ = ensure!(true);
   |                         ^ missing tokens in macro arguments
   |
note: while trying to match `,`
  --> src/macros.rs
   |
   |     ($cond:expr, $err:expr $(,)?) => {{
   |                ^
